src/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonValueConverterEnum.cs

```diff
@@ -2,168 +2,275 @@
+                }
+
+                if (IsValidIdentifier(original))
+                {
+                    transformed = _namingPolicy.ConvertName(original);
+                    writer.WriteStringValue(transformed);
+                    if (_nameCache != null)
+                    {
+                        _nameCache.Add(original, transformed);
+                    }
+                    return;
+                }
             }
-            else if (s_enumTypeCode == TypeCode.UInt64)
+
+            if (!_converterOptions.HasFlag(EnumConverterOptions.AllowNumbers))
             {
-                // Use the ulong converter to prevent conversion into a signed\long value.
-                ulong ulongValue = Convert.ToUInt64(value);
-                writer.WriteNumberValue(ulongValue);
+                ThrowHelper.ThrowFormatException();
             }
-            else
+
+            switch (s_enumTypeCode)
             {
-                // long can hold the signed\unsigned values of other integer types
-                long longValue = Convert.ToInt64(value);
-                writer.WriteNumberValue(longValue);
+                case TypeCode.Int32:
+                    writer.WriteNumberValue(Unsafe.As<T, int>(ref value));
+                    break;
+                case TypeCode.UInt32:
+                    writer.WriteNumberValue(Unsafe.As<T, uint>(ref value));
+                    break;
+                case TypeCode.UInt64:
+                    writer.WriteNumberValue(Unsafe.As<T, ulong>(ref value));
+                    break;
+                case TypeCode.Int64:
+                    writer.WriteNumberValue(Unsafe.As<T, long>(ref value));
+                    break;
+                case TypeCode.Int16:
+                    writer.WriteNumberValue(Unsafe.As<T, short>(ref value));
+                    break;
+                case TypeCode.UInt16:
+                    writer.WriteNumberValue(Unsafe.As<T, ushort>(ref value));
+                    break;
+                case TypeCode.Byte:
+                    writer.WriteNumberValue(Unsafe.As<T, byte>(ref value));
+                    break;
+                case TypeCode.SByte:
+                    writer.WriteNumberValue(Unsafe.As<T, sbyte>(ref value));
+                    break;
+                default:
+                    ThrowHelper.ThrowFormatException();
+                    break;
             }
         }
 
         public override void Write(Utf8JsonWriter writer, T value, JsonEncodedText propertyName, JsonSerializerOptions options)
         {
-            if (TreatAsString)
+            // If strings are allowed, attempt to write it out as a string value
+            if (_converterOptions.HasFlag(EnumConverterOptions.AllowStrings))
             {
-                writer.WriteString(propertyName, value.ToString());
+                string original = value.ToString();
+                if (_nameCache != null && _nameCache.TryGetValue(original, out string transformed))
```

Comment:
What is the lifetime of a converter? How long will this cache live for?

Reply:
As long as JsonSerializerOptions it was attached to.

Reply:
In that case you need to worry about thread-safety. Use ConcurrentDictionary

Reply:
Ok.

Reply:
Alternatively you could eagerly resolve the enum names at once and set to a dictionary. That would avoid locking every TryGetValue.